templates/instanceha/bin/instanceha.py

```diff
@@ -498,6 +498,33 @@ def _redfish_reset(url, user, passwd, timeout, action):
     r = requests.post(url, data=json.dumps(payload), headers=headers, auth=(user, passwd), verify=False, timeout=timeout)
     return r
 
+def _bmh_fence(token, namespace, host, action):
+
+    url = "https://kubernetes.default.svc/apis/metal3.io/v1alpha1/namespaces/%s/baremetalhosts/%s?fieldManager=kubectl-patch" % (namespace, host)
+    headers={'Authorization': 'Bearer '+token, 'Content-Type': 'application/merge-patch+json'}
+    cacert = '/var/run/secrets/kubernetes.io/serviceaccount/ca.crt'
+
+    if action == 'off':
+        ann={"metadata":{"annotations":{"reboot.metal3.io/iha":"{\"mode\": \"hard\"}"}}}
+        r = requests.patch(url, headers=headers, verify=cacert, data=json.dumps(ann))
+        if r.status_code == 200:
+
+            #check if server is off - wait up to 30s - TODO this may need tuning
+            headers={'Authorization': 'Bearer '+token}
+            url = "https://kubernetes.default.svc/apis/metal3.io/v1alpha1/namespaces/%s/baremetalhosts/%s" % (namespace, host)
+
+            timeout_at = time.time() + 30
+            while time.time() < timeout_at:
+              time.sleep(3)
+              s = requests.get(url, headers=headers, verify=cacert)
+              poweredon = json.loads(s.text)['status']['poweredOn']
+              if not poweredon:
+                break
+            return not poweredon
+    else:
+        ann={"metadata":{"annotations":{"reboot.metal3.io/iha":None}}}
```

Comment:
The docs [1] seem to indicate that the annotation has to be remove completely for the BMH to be powered back on.  It seems like this is just removing the mode but otherwise keeping the annotation.

[1] https://book.metal3.io/bmo/reboot_annotation.html#phased-reboot

Reply:
That's what it might look like but from my tests the annotation is removed completely.

Reply:
I am rebuilding the env, will attach the results as soon as I have them.

Reply:
_bmh_fence(token, namespace, 'edpm-compute-0', 'off')

```
[cifmw@haa-09 watcher-operator]$ oc get bmh edpm-compute-0 -o json |jq .metadata.annotations
{
  "FOO": "BAR",
  "inspect.metal3.io": "disabled",
  "kubectl.kubernetes.io/last-applied-configuration": "{\"apiVersion\":\"metal3.io/v1alpha1\",\"kind\":\"BareMetalHost\",\"metadata\":{\"annotations\":{\"inspect.metal3.io\":\"disabled\"},\"labels\":{\"app\":\"openstack\",\"created-by\":\"install_yamls\"},\"name\":\"edpm-compute-0\",\"namespace\":\"openstack\"},\"spec\":{\"bmc\":{\"address\":\"redfish-virtualmedia+http://sushy-emulator.apps-crc.testing/redfish/v1/Systems/96174cdf-9282-408e-99f9-0a4fadfd2b8b\",\"credentialsName\":\"node-0-bmc-secret\"},\"bootMACAddress\":\"52:54:00:3f:9c:1a\",\"bootMode\":\"UEFI\",\"online\":false,\"rootDeviceHints\":{\"deviceName\":\"/dev/vda\"}}}\n",
  "reboot.metal3.io/iha": "{\"mode\": \"hard\"}"
}
```
_bmh_fence(token, namespace, 'edpm-compute-0', 'on')

```
[cifmw@haa-09 watcher-operator]$ oc get bmh edpm-compute-0 -o json |jq .metadata.annotations
{
  "FOO": "BAR",
  "inspect.metal3.io": "disabled",
  "kubectl.kubernetes.io/last-applied-configuration": "{\"apiVersion\":\"metal3.io/v1alpha1\",\"kind\":\"BareMetalHost\",\"metadata\":{\"annotations\":{\"inspect.metal3.io\":\"disabled\"},\"labels\":{\"app\":\"openstack\",\"created-by\":\"install_yamls\"},\"name\":\"edpm-compute-0\",\"namespace\":\"openstack\"},\"spec\":{\"bmc\":{\"address\":\"redfish-virtualmedia+http://sushy-emulator.apps-crc.testing/redfish/v1/Systems/96174cdf-9282-408e-99f9-0a4fadfd2b8b\",\"credentialsName\":\"node-0-bmc-secret\"},\"bootMACAddress\":\"52:54:00:3f:9c:1a\",\"bootMode\":\"UEFI\",\"online\":false,\"rootDeviceHints\":{\"deviceName\":\"/dev/vda\"}}}\n"
}
```

Reply:
Looks good :+1: